File: src/1619_Pond_Sizes_LCCI.py
# 1619_Pond_Sizes_LCCI
# https://leetcode.cn/problems/pond-sizes-lcci/

class Solution:
    def pondSizes(self, land: List[List[int]]) -> List[int]:
        m, n = len(land), len(land[0])

        def dfs(x: int, y: int) -> int:
            if x < 0 or x >= m or y < 0 or y >= n or land[x][y] != 0:
                return 0
            
            land[x][y] = -1
            res = 1
            for dx in [-1, 0, 1]:
                for dy in [-1, 0, 1]:
                    if dx == dy == 0:
                        continue
                    res += dfs(x + dx, y + dy)
            return res
        
        res = list()
        for i in range(m):
            for j in range(n):
                if land[i][j] == 0:
                    res.append(dfs(i, j))
        res.sort()
        return res

                

        
class Solution:
    def pondSizes(self, land: List[List[int]]) -> List[int]:
        m, n = len(land), len(land[0])
        
        def bfs(x: int, y: int) -> int:
            res = 0
            q = deque([(x, y)])
            land[x][y] = -1

            while q:
                x, y = q.popleft()
                res += 1
                for dx in [-1, 0, 1]:
                    for dy in [-1, 0, 1]:
                        if dx == dy == 0:
                            continue
                        if x + dx < 0 or x + dx >= m or y + dy < 0 or y + dy >= n or land[x + dx][y + dy] != 0:
                            continue
                        land[x + dx][y + dy] = -1
                        q.append((x + dx, y + dy))
            return res
        
        res = list()
        for i in range(m):
            for j in range(n):
                if land[i][j] == 0:
                    res.append(bfs(i, j))
        res.sort()
        return res

File: src/1619_Pond_Sizes_LCCI.rs
// 面试题 16.19. 水域大小
// https://leetcode.cn/problems/pond-sizes-lcci/

impl Solution {
    pub fn pond_sizes(mut land: Vec<Vec<i32>>) -> Vec<i32> {
        let m = land.len();
        let n = land[0].len();

        fn dfs(x: usize, y: usize, land: &mut Vec<Vec<i32>>) -> i32 {
            let m = land.len();
            let n = land[0].len();

            if x >= m || y >= n || land[x][y] != 0 {
                return 0;
            }

            land[x][y] = -1;
            let mut res = 1;

            for dx in -1..=1 {
                for dy in -1..=1 {
                    if dx == 0 && dy == 0 {
                        continue;
                    }

                    let nx = x as i32 + dx;
                    let ny = y as i32 + dy;

                    if nx >= 0 && ny >= 0 {
                        res += dfs(nx as usize, ny as usize, land);
                    }
                }
            }

            res
        }

        let mut res = Vec::new();
        for i in 0..m {
            for j in 0..n {
                if land[i][j] == 0 {
                    res.push(dfs(i, j, &mut land));
                }
            }
        }

        res.sort();
        res
    }
}